### ISSUE TITLE
Segfault when forwarder list contains local server itself

### ISSUE THREAD

**Issue by hhoffstaette** [coredns/coredns#2161]
I recently migrated from dnsmasq and very much like coredns overall (now running everywhere),
but of course also immediately found a bug. :)

With the following config:
```
. {
	forward . /etc/resolv.conf
	cache
	log
}
```
and the following resolv.conf:
```
# host-local instance for resilience/caching
nameserver 127.0.0.1
# LAN-local instance that forwards to my ISP etc.
nameserver 192.168.100.222
```
coredns repeatably crashes with the follwing log a few seconds after start & a bit of initial random
lookup activity:
```
coredns -conf crash.conf
.:53
2018/10/05 01:25:11 [INFO] CoreDNS-1.2.2
2018/10/05 01:25:11 [INFO] linux/amd64, go1.11, eb51e8bac90fac86d34c9e1cb89b04ea0936b034
CoreDNS-1.2.2
linux/amd64, go1.11, eb51e8bac90fac86d34c9e1cb89b04ea0936b034
127.0.0.1:56626 - [05/Oct/2018:01:25:15 +0200] 35427 "PTR IN 166.241.155.62.in-addr.arpa. udp 45 false 512" NXDOMAIN qr,rd,ra 139 0.017405332s
127.0.0.1:50454 - [05/Oct/2018:01:25:15 +0200] 35427 "PTR IN 166.241.155.62.in-addr.arpa. udp 45 false 512" NXDOMAIN qr,rd,ra 139 0.01795297s
127.0.0.1:56209 - [05/Oct/2018:01:25:15 +0200] 22338 "PTR IN 38.118.5.217.in-addr.arpa. udp 43 false 512" NXDOMAIN qr,rd,ra 140 0.000773091s
127.0.0.1:51105 - [05/Oct/2018:01:25:24 +0200] 57904 "AAAA IN ntp3.ptb.de. udp 29 false 512" NOERROR qr,rd,ra 112 0.000827708s
127.0.0.1:58737 - [05/Oct/2018:01:25:24 +0200] 61978 "A IN ntp3.ptb.de. udp 29 false 512" NOERROR qr,rd,ra 100 0.000752208s
127.0.0.1:56626 - [05/Oct/2018:01:25:24 +0200] 61978 "A IN ntp3.ptb.de. udp 29 false 512" NOERROR qr,rd,ra 100 0.001321987s
127.0.0.1:51105 - [05/Oct/2018:01:25:24 +0200] 61978 "A IN ntp3.ptb.de. udp 29 false 512" NOERROR qr,rd,ra 100 0.001900355s
127.0.0.1:60269 - [05/Oct/2018:01:25:25 +0200] 63069 "PTR IN 3.0.1.0.0.0.0.0.0.0.0.0.0.0.0.0.1.0.e.b.0.1.6.0.8.3.6.0.1.0.0.2.ip6.arpa. udp 90 false 512" NOERROR qr,rd,ra 191 0.000858299s
127.0.0.1:59572 - [05/Oct/2018:01:25:41 +0200] 29616 "AAAA IN time1.google.com. udp 34 false 512" NOERROR qr,rd,ra 78 0.009314037s
127.0.0.1:56671 - [05/Oct/2018:01:25:41 +0200] 64920 "A IN time1.google.com. udp 34 false 512" NOERROR qr,rd,ra 66 0.022090466s
127.0.0.1:59572 - [05/Oct/2018:01:25:41 +0200] 64920 "A IN time1.google.com. udp 34 false 512" NOERROR qr,rd,ra 66 0.022850007s
127.0.0.1:56671 - [05/Oct/2018:01:25:42 +0200] 53596 "PTR IN 2.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.f.0.0.8.4.0.3.1.0.0.0.0.3.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 181 0.068109175s
127.0.0.1:42052 - [05/Oct/2018:01:25:42 +0200] 53596 "PTR IN 2.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.f.0.0.8.4.0.3.1.0.0.0.0.3.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 181 0.068753824s
127.0.0.1:58481 - [05/Oct/2018:01:25:42 +0200] 10822 "PTR IN 1.0.0.0.1.0.0.0.0.0.0.0.0.0.0.0.8.4.0.0.0.0.0.8.0.5.4.1.0.0.a.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 184 0.027052733s
127.0.0.1:56671 - [05/Oct/2018:01:25:42 +0200] 10822 "PTR IN 1.0.0.0.1.0.0.0.0.0.0.0.0.0.0.0.8.4.0.0.0.0.0.8.0.5.4.1.0.0.a.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 184 0.027765411s
127.0.0.1:36055 - [05/Oct/2018:01:25:42 +0200] 10822 "PTR IN 1.0.0.0.1.0.0.0.0.0.0.0.0.0.0.0.8.4.0.0.0.0.0.8.0.5.4.1.0.0.a.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 184 0.028311023s
127.0.0.1:50097 - [05/Oct/2018:01:25:42 +0200] 20716 "PTR IN c.6.1.2.0.0.0.0.0.0.0.0.0.0.0.0.1.0.0.0.0.0.0.0.0.6.8.4.1.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 192 0.042064483s
127.0.0.1:44948 - [05/Oct/2018:01:25:42 +0200] 23716 "PTR IN 1.1.0.0.0.0.0.0.0.0.0.0.0.0.0.0.f.d.1.1.0.0.0.0.0.6.8.4.1.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 192 0.036674568s
127.0.0.1:58481 - [05/Oct/2018:01:25:42 +0200] 60421 "PTR IN 4.7.8.f.0.0.0.4.c.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.8.4.1.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 192 0.029485152s
127.0.0.1:56671 - [05/Oct/2018:01:25:42 +0200] 60421 "PTR IN 4.7.8.f.0.0.0.4.c.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.8.4.1.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 192 0.030067642s
127.0.0.1:47395 - [05/Oct/2018:01:25:42 +0200] 60421 "PTR IN 4.7.8.f.0.0.0.4.c.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.8.4.1.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 192 0.030589227s
127.0.0.1:56061 - [05/Oct/2018:01:25:42 +0200] 49242 "PTR IN 1.4.e.c.0.0.0.4.c.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.8.4.1.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 192 0.02515867s
127.0.0.1:50750 - [05/Oct/2018:01:25:42 +0200] 26028 "PTR IN 8.8.3.d.0.0.0.4.c.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.8.4.1.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 192 0.022117572s
127.0.0.1:60533 - [05/Oct/2018:01:25:43 +0200] 54255 "PTR IN 9.0.1.d.0.0.0.0.c.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.8.4.1.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 192 0.125546181s
127.0.0.1:49160 - [05/Oct/2018:01:25:43 +0200] 6629 "PTR IN 2.6.3.e.0.0.0.4.c.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.0.6.8.4.1.0.0.2.ip6.arpa. udp 90 false 512" NXDOMAIN qr,rd,ra 192 0.117271752s
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x11e3939]

goroutine 42 [running]:
github.com/coredns/coredns/plugin/forward.(*Proxy).Healthcheck.func1(0x4363f9, 0x1938450)
	/tmp/portage/net-dns/coredns-1.2.2/work/coredns-1.2.2/src/github.com/coredns/coredns/plugin/forward/proxy.go:54 +0x29
github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do.func1(0xc0004a86b0, 0x1dcd6500, 0xc00014a5c0)
	/tmp/portage/net-dns/coredns-1.2.2/work/coredns-1.2.2/src/github.com/coredns/coredns/plugin/pkg/up/up.go:38 +0x36
created by github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do
	/tmp/portage/net-dns/coredns-1.2.2/work/coredns-1.2.2/src/github.com/coredns/coredns/plugin/pkg/up/up.go:36 +0xa0
```

This happens both with my local build (Gentoo x86_64, Go 1.11) as well as the prebuilt binaries,
both 1.2.0 and 1.2.2. I don't know much golang so I'm afraid the stack trace above is all I can provide;
however I *can* test & verify patches as long as they easily apply to 1.2.2.
I understand of course that the scenario of forwarding to yourself is strictly nonsense, but IMHO
it should nevertheless work (just like dnsmasq), e.g. by fixing the forwarder plugin to filter out its
own server, or - as worst case - failing. Using resolv.conf as above was simply the first thing
I tried as initial attempt, and of course found the bug; I now manually specify a forwarder list - which
no longer contains the local instance - and all is well.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[coredns/coredns] Segfault when for..." ]
>I recently migrated from dnsmasq and very much like coredns overall (now running everywhere),
>but of course also immediately found a bug. :)
>
>With the following config:
>```
>. {
>	forward . /etc/resolv.conf
>	cache
>	log
>}
>```
>and the following resolv.conf:
>```
># host-local instance for resilience/caching
>nameserver 127.0.0.1
># LAN-local instance that forwards to my ISP etc.
>nameserver 192.168.100.222
>```

Think you just created a loop. Can you load the *loop* plugin and see if coredns 
stops itself because it detected a loop?

**Comment by hhoffstaette** [comment]
Added the loop plugin but no luck - same kaboom, this time even in two goroutines at the same time:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x11e3939]

goroutine 115 [running]:
github.com/coredns/coredns/plugin/forward.(*Proxy).Healthcheck.func1(0x0, 0x0)
	/tmp/portage/net-dns/coredns-1.2.2/work/coredns-1.2.2/src/github.com/coredns/coredns/plugin/forward/proxy.go:54 +0x29
github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do.func1(0xc000428190, 0x1dcd6500, 0xc00044c5a0)
	/tmp/portage/net-dns/coredns-1.2.2/work/coredns-1.2.2/src/github.com/coredns/coredns/plugin/pkg/up/up.go:38 +0x36
created by github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do
	/tmp/portage/net-dns/coredns-1.2.2/work/coredns-1.2.2/src/github.com/coredns/coredns/plugin/pkg/up/up.go:36 +0xa0
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x11e3939]

goroutine 68 [running]:
github.com/coredns/coredns/plugin/forward.(*Proxy).Healthcheck.func1(0x4363f9, 0x1938450)
	/tmp/portage/net-dns/coredns-1.2.2/work/coredns-1.2.2/src/github.com/coredns/coredns/plugin/forward/proxy.go:54 +0x29
github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do.func1(0xc0004c4130, 0x1dcd6500, 0xc00044c5c0)
	/tmp/portage/net-dns/coredns-1.2.2/work/coredns-1.2.2/src/github.com/coredns/coredns/plugin/pkg/up/up.go:38 +0x36
created by github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do
	/tmp/portage/net-dns/coredns-1.2.2/work/coredns-1.2.2/src/github.com/coredns/coredns/plugin/pkg/up/up.go:36 +0xa0
```

**Comment by rdrozhdzh** [comment]
@hhoffstaette, can you just use
```
. {
	forward . 192.168.100.222
	cache
	log
}
```
or it is unacceptable to you?

**Comment by hhoffstaette** [comment]
> @hhoffstaette, can you just use
> 	forward . 192.168.100.222
> or it is unacceptable to you?

I already wrote that this (with more forwarders etc.) is what I already am using (and it works fine),
but that is not the point. The problems are loop detection & especially the runtime crash, which
should never happen.

**Comment by rdrozhdzh** [comment]
Oh, I see, at least we know for sure that the problem is related to loop detection

**Comment by chrisohaver** [comment]
When I test locally (with loop plugin enabled) ...

If I include only one nameserver (127.0.0.1) in resolv.conf, the loop detection catches the loop every time.
If I include two nameservers (as in the example), Things just work.  Presumably 127.0.0.1 is marked as unhealthy, and requsts are forwarded to the second nameserver.

**Comment by chrisohaver** [comment]
> Oh, I see, at least we know for sure that the problem is related to loop detection

But it crashes without the loop plugin enabled.

**Comment by rdrozhdzh** [comment]
I mean, the problem definitely exists, and now we better understand which direction to dig

**Comment by chrisohaver** [comment]
I double checked - I cannot reproduce the crash with the `loop` plugin enabled.  In fact, it seems that health checking wins, and disables the 127.0.0.1 upstream before the loop detection can detect it. So, in my tests at least, it seems to be working as expected.  

I am running on a local build @ head.

Without `loop` enabled, it also works... so something is different here in our setups.  Maybe something special about the local LAN dns server coming into play.    I am using `8.8.8.8` as the second upstream server.  Does `forward` follow the order of nameservers in the `/etc/resolv.conf` file in the order they are listed?

**Comment by rdrozhdzh** [comment]
> Does forward follow the order of nameservers in the /etc/resolv.conf file in the order they are listed?

By default forward uses `random` policy, so on each incoming request it randomly chooses upstream server

**Comment by rdrozhdzh** [comment]
@hhoffstaette , did you try coredns v1.1.4? Just curious if it works fine or not

**Comment by rdrozhdzh** [comment]
@hhoffstaette , could you cherry-pick  the change from #2165 and check if it fixes your problem?

**Comment by hhoffstaette** [comment]
> @hhoffstaette , could you cherry-pick the change from #2165 and check if it fixes your problem?

The patch alone does not compile due to previous changes (no "trans" variable), but I then
cherry-picked the preceding commit c349446a2344 ("Cleanup ParseHostOrFile") as well and
with those two it no longer crashes. Happily running in the background doing its thing, health-
checking and logging and whatnot.

**Comment by hhoffstaette** [comment]
> @hhoffstaette , did you try coredns v1.1.4? Just curious if it works fine or not

Never tried 1.1.x and won't since moving forward makes more sense. :)

**Comment by rdrozhdzh** [comment]
thank you @hhoffstaette for testing. 
I was asking about v1.1.4 because I suppose the regression might appear in 1.2.0
Please let us know if you'll see some warnings like 'No healthckecker' in coredns log with a patch from #2165 applied.

**Comment by kometchtech** [comment]
I also have an event that seems to be the same problem.

```
2018/10/14 04:54:34 [INFO] Reloading complete
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xbc7260]

goroutine 169 [running]:
github.com/coredns/coredns/plugin/forward.(*Proxy).Healthcheck.func1(0x0, 0x66058)
        /tmp/tmp.onHxXAXP70/g/src/github.com/coredns/coredns/plugin/forward/proxy.go:54 +0x20
github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do.func1(0x4000450d50, 0x0, 0x400050b4c0)
        /tmp/tmp.onHxXAXP70/g/src/github.com/coredns/coredns/plugin/pkg/up/up.go:38 +0x2c
created by github.com/coredns/coredns/plugin/pkg/up.(*Probe).Do
        /tmp/tmp.onHxXAXP70/g/src/github.com/coredns/coredns/plugin/pkg/up/up.go:36 +0x88
.:53
2018/10/14 05:07:40 [INFO] CoreDNS-1.2.2
2018/10/14 05:07:40 [INFO] linux/arm64, go1.11, eb51e8b
```

**Comment by hhoffstaette** [comment]
This is fixed since 1.2.3 (thanks!)  - what's the procedure here? Am I supposed to close it?
